File: python/rational-numbers/1/rational_numbers.py
import math


class Rational:
    def __init__(self, numer, denom):
        self.numer = numer
        self.denom = denom
        self.lowest_terms()

    def __eq__(self, other):
        return self.numer == other.numer and self.denom == other.denom

    def __repr__(self):
        return f'{self.numer}/{self.denom}'

    def __add__(self, other):
        self.numer = self.numer * other.denom + other.numer * self.denom
        self.denom = self.denom * other.denom
        self.lowest_terms()
        return self

    def __sub__(self, other):
        self.numer = self.numer * other.denom - other.numer * self.denom
        self.denom = self.denom * other.denom
        self.lowest_terms()
        return self

    def __mul__(self, other):
        self.numer = self.numer * other.numer
        self.denom = self.denom * other.denom
        self.lowest_terms()
        return self

    def __truediv__(self, other):
        self.numer = self.numer * other.denom
        self.denom = self.denom * other.numer
        self.lowest_terms()
        return self

    def __abs__(self):
        self.numer = abs(self.numer)
        self.denom = abs(self.denom)
        self.lowest_terms()
        return self

    def __pow__(self, power):
        if isinstance(power, int):
            if power >= 0:
                self.numer = self.numer ** power
                self.denom = self.denom ** power
            else:
                self.numer, self.denom = (self.denom ** -power, self.numer ** -power)
        else:
            # real power returns real
            return self.numer ** power / self.denom ** power
        self.lowest_terms()
        return self

    def __rpow__(self, base):
        return (base ** self.numer) ** (1 / self.denom)

    def lowest_terms(self):
        hcf = math.gcd(self.numer, self.denom)
        self.numer //= hcf
        self.denom //= hcf
        if self.denom< 0:
            self.numer *= -1
            self.denom *= -1

File: julia/rational-numbers/1/rational-numbers.jl
import Base: +, *, ^, -, /, <, <=, promote_rule

struct RationalNumber{T<:Integer} <: Real
    numer::T
    denom::T

    function RationalNumber{T}(n, d) where T<:Integer 
        (n ≠ 0 || d ≠ 0) || throw(ArgumentError("0/0 is invalid"))
        hcf = gcd(n, d)
        num, den = hcf > 0 ? (n ÷ hcf, d ÷ hcf) : (n, d)
        d ≥ 0 ? new(num, den) : new(-num, -den)
    end
end

RationalNumber(num::T, den::T) where {T<:Integer} = RationalNumber{T}(num, den)

RationalNumber{T}(r1::Integer) where {T<:Integer} = RationalNumber{T}(convert(T, r1), convert(T, 1))

promote_rule(::Type{RationalNumber{T}}, ::Type{S}) where {T<:Integer,S<:Integer} = 
    RationalNumber{promote_type(T,S)}

numerator(r::RationalNumber) = r.numer
denominator(r::RationalNumber) = r.denom

abs(rn::RationalNumber) = RationalNumber(Base.abs(rn.numer), Base.abs(rn.denom))

+(r1::RationalNumber, r2::RationalNumber) = 
    RationalNumber(r1.numer * r2.denom + r2.numer * r1.denom, r1.denom * r2.denom)

-(r1::RationalNumber, r2::RationalNumber) = 
    RationalNumber(r1.numer * r2.denom - r2.numer * r1.denom, r1.denom * r2.denom)

*(r1::RationalNumber, r2::RationalNumber) = RationalNumber(r1.numer * r2.numer, r1.denom * r2.denom)

/(r1::RationalNumber, r2::RationalNumber) = RationalNumber(r1.numer * r2.denom, r1.denom * r2.numer)

^(base::Real, r1::RationalNumber) = (float(base)^r1.numer)^(1 / r1.denom)

<(r1::RationalNumber, r2::RationalNumber) = 
    r1.denom == r2.denom ? r1.numer < r2.numer : r1.numer * r2.denom < r2.numer * r1.denom

<=(r1::RationalNumber, r2::RationalNumber) = 
    r1.denom == r2.denom ? r1.numer <= r2.numer : r1.numer * r2.denom <= r2.numer * r1.denom
